 (\(\
 (-.-)
o('')('')

# Emend Alignment of Spliced Transcript Reads (EASTR)
EASTR is a software tool that finds and removes spurious spliced alignments caused by repeated sequences. 
EASTR can be applied to any RNA-seq dataset regardless of the alignment software used.

<!-- TODO: Give a quick sentence or two to explain what this should do/give you. -->
## Required Dependencies

- [bowtie2](https://github.com/BenLangmead/bowtie2)
- [samtools](https://github.com/samtools/samtools)

## Getting Started

1. Clone source repository

	```bash
	git clone https://github.com/ishinder/EASTR
	cd EASTR
	```

2. Install EASTR
	```bash
	# (OPTIONAL) Install in a Python virtual environment
	# python3 -m virtualenv venv # (OPTIONAL)
	# source .venv/bin/activate # (OPTIONAL)
	make install # Install EASTR package
	```

3. OPTIONAL: Download test data
  ```bash
  wget ftp://ftp.ccb.jhu.edu/pub/ishinder/chrX.gtf
  ```

## Usage
```shell
usage: EASTR [-h] [-R REFERENCE] [-bam BAM] [-A A] [-B B] [-O O O] [-E E E] [-k K] [--scoreN SCOREN] [-w W] [-p P] [-o FILE]

Emend alignments of spuriously spliced transcript reads

options:
  -h, --help            show this help message and exit
  -R REFERENCE, --reference REFERENCE
                        reference fasta genome used in alignment
  -bam BAM              Input BAM file to emend alignments
  -A A                  Matching score, default = 2
  -B B                  Mismatching penalty, default = 4
  -O O O                Gap open penalty, default = [4, 24]
  -E E E                Gap extension penalty, default = [2, 1]. A gap of length k costs min(O1+k*E1, O2+k*E2).
  -k K                  kmer length for alignment, default=7
  --scoreN SCOREN       Score of a mismatch involving ambiguous bases, default=1
  -w W                  minimizer window size, default=7
  -p P                  Number of parallel processes, default=1
  -o FILE               write output to FILE; the default output is to terminal
```

## Input files

### Alignment file
The input file must be a BAM file generated by a splice aware aligner (HISAT2 or STAR)
The alignment file should be sorted by coordinate.

### Reference fasta file 
The reference fasta file is used to obtain flanking intronic sequences. 
This file should match the contigs in the provided BAM file.

### Bowtie2 index 


### Reference annotation file



## Running a test file


<!-- TODO: Explain what output you are expected to see and brief explanation -->
```shell
eastr -R tests/data/chrX.fa \
    -bam tests/data/ERR188044_chrX.bam \
    --out_introns tests/output/ERR188044_chrX_spurious_introns.bed
    --out_bam tests/output/ERR188044_chrX_filtered.bam
    --removed_reads tests/output/ERR188044_chrX_removed_reads.txt
```

# Citation
To cite EASTR in publications, please use:
[citation]
